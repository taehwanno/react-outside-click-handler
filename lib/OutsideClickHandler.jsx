import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { addEventListener } from 'consolidated-events';

const propTypes = {
  children: PropTypes.node,
  onOutsideClick: PropTypes.func,
  useCapture: PropTypes.bool,
};

const defaultProps = {
  children: <span />,
  onOutsideClick() {},
  useCapture: true,
};

export default class OutsideClickHandler extends Component {
  constructor(...args) {
    super(...args);

    this.onOutsideClick = this.onOutsideClick.bind(this);
    this.setChildNodeRef = this.setChildNodeRef.bind(this);
  }

  componentDidMount() {
    const { useCapture } = this.props;
    this.removeEventListener = addEventListener(document, 'click', this.onOutsideClick, { capture: useCapture });
  }

  componentWillUnmount() {
    if (this.removeEventListener) {
      this.removeEventListener();
    }
  }

  onOutsideClick(e) {
    const { onOutsideClick } = this.props;
    const { childNode } = this;
    const isDescendantOfRoot = childNode && childNode.contains(e.target);
    if (!isDescendantOfRoot) {
      onOutsideClick(e);
    }
  }

  setChildNodeRef(ref) {
    this.childNode = ref;
  }

  render() {
    const { children } = this.props;
    return <div ref={this.setChildNodeRef}>{children}</div>;
  }
}

OutsideClickHandler.propTypes = propTypes;
OutsideClickHandler.defaultProps = defaultProps;
