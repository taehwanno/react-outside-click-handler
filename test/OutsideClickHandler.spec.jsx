import React from 'react';
import { shallow } from 'enzyme';

import OutsideClickHandler from '../lib/OutsideClickHandler';

describe('<OutsideClickHandler />', () => {
  describe('#render', () => {
    it('should match to snapshot when render by default', () => {
      const wrapper = shallow(<OutsideClickHandler />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('#componentDidMount', () => {
    it('should add click event listener to document', () => {
      const addEventListener = jest.spyOn(document, 'addEventListener');
      const wrapper = shallow(<OutsideClickHandler />);
      expect(addEventListener).toHaveBeenCalledTimes(1);
      expect(addEventListener).toHaveBeenCalledWith('click', wrapper.instance().handleOutsideClick, true);
      addEventListener.mockRestore();
    });
  });

  describe('#componentWillUnmount', () => {
    it('should remove click event listener to document', () => {
      const removeEventListener = jest.spyOn(document, 'removeEventListener');
      const wrapper = shallow(<OutsideClickHandler />);
      const { handleOutsideClick } = wrapper.instance();
      wrapper.unmount();
      expect(removeEventListener).toHaveBeenCalledTimes(1);
      expect(removeEventListener).toHaveBeenCalledWith('click', handleOutsideClick, true);
      removeEventListener.mockRestore();
    });
  });

  describe('#setChildNode', () => {
    it('should set childNode', () => {
      const ref = { value: 1 };
      const wrapper = shallow(<OutsideClickHandler />);
      wrapper.instance().setChildNode(ref);
      expect(wrapper.instance().childNode).toEqual(ref);
    });
  });

  describe('#handleOutsideClick', () => {
    it('should calls props.onOutsideClick when event target is not decendant of this.childNode', () => {
      const event = { target: null };
      const onOutsideClick = jest.fn();
      const wrapper = shallow(<OutsideClickHandler onOutsideClick={onOutsideClick} />);
      wrapper.instance().childNode = { contains() { return false; } };
      wrapper.instance().handleOutsideClick(event);
      expect(onOutsideClick).toHaveBeenCalledWith(event);
    });

    it('should not calls props.onOutsideClick when event target is decendant of this.childNode', () => {
      const event = { target: null };
      const onOutsideClick = jest.fn();
      const wrapper = shallow(<OutsideClickHandler onOutsideClick={onOutsideClick} />);
      wrapper.instance().childNode = { contains() { return true; } };
      wrapper.instance().handleOutsideClick(event);
      expect(onOutsideClick).not.toHaveBeenCalled();
    });
  });
});
