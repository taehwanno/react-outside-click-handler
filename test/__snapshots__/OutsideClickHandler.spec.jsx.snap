// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OutsideClickHandler /> #render should match to snapshot when render by default 1`] = `
<div>
  <span />
</div>
`;
